src/Riok.Mapperly/SyntaxProvider.Roslyn3.9.cs

```diff
@@ -0,0 +1,46 @@
+﻿#if ROSLYN3_9_OR_GREATER && !ROSLYN4_0_OR_GREATER
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Riok.Mapperly;
+
+class SyntaxProvider : ISyntaxContextReceiver
+{
+    public List<ClassDeclarationSyntax> ClassDeclarations { get; } = new();
+
+    public void OnVisitSyntaxNode(GeneratorSyntaxContext context)
+    {
+        if (IsSyntaxTargetForGeneration(context.Node))
```

Comment:
To keep the nesting level low and improve readability an inverted if with an early return could help here.